File: video_47_Number_of_Longest_Increasing_Subsequence.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution{
    public :
    int NumberofLIS(int n, vector<int>&nums) {
        // Code here
        vector<int> dp(n, 1), count(n, 1);
        int maxi = 1;
        for(int i = 0; i<n; i++){
            for(int j = 0; j<i; j++){
                if(nums[i]>nums[j] && 1+dp[j]>dp[i]){
                    dp[i] = 1+dp[j];
                    count[i] = count[j];
                }
                else if(nums[i]>nums[j] && 1+dp[j] == dp[i])
                    count[i]+=count[j];
                maxi = max(maxi, dp[i]);
            }
        }
        
        int ans = 0;
        for(int i = 0; i<n; i++)
        if(dp[i] == maxi) ans+=count[i];

        return ans;
    }
};

int main(){
    int n; cin>>n;
    vector<int> arr(n);
    for(int i = 0; i<n; i++)
    cin>>arr[i];

    Solution obj;
    cout<<obj.NumberofLIS(n, arr);
    return 0;
}

File: video_47_Number_of_Longest_Increasing_Subsequence.java
import java.util.*;
class video_47_Number_of_Longest_Increasing_Subsequence{
    public int NumberofLIS(int N, int arr[]){
        // Code here
        int dp[] = new int[N], count[] = new int[N];
        Arrays.fill(dp, 1);
        Arrays.fill(count, 1);
        int maxLen = 1;
        for(int i = 0; i<N; i++){
            for(int j = 0; j<i; j++){
                if(arr[i]>arr[j] && 1+dp[j]>dp[i]){
                    dp[i] = 1+dp[j];
                    count[i] = count[j];
                }
                else if(arr[i]>arr[j] && dp[i] == 1+dp[j])
                    count[i]+=count[j];
                maxLen = Math.max(maxLen, dp[i]);
            }
        }

        int ans = 0;
        for(int i = 0; i<N; i++)
        if(dp[i] == maxLen) ans+=count[i]; 
        return ans;
    }
}